### PR TITLE
Add additional permission checks in AuditMessageIT

### PR DESCRIPTION
**Pull request by Manno15**
In relation to #1830, I only added an audit check for flushing a table at the moment but I am working on adding additional ones for set/remove system property and any other that is requested.

### CODE REVIEW

File: test/src/main/java/org/apache/accumulo/test/AuditMessageIT.java [hunk 1]
```diff
@@ -482,6 +482,10 @@ public void testDeniedAudits() throws AccumuloSecurityException, AccumuloExcepti
       auditConnector.tableOperations().deleteRows(OLD_TEST_TABLE_NAME, new Text("myRow"),
           new Text("myRow~"));
     } catch (AccumuloSecurityException ex) {}
+    try {
+      auditConnector.tableOperations().flush(OLD_TEST_TABLE_NAME, new Text("start"),
+              new Text("end"), false);
```
Comment (ctubbsii) [inline test/src/main/java/org/apache/accumulo/test/AuditMessageIT.java:487]: Normally, a range would be specified so that the start sorts before the end, but 's' sorts after 'e'. You should change it to match `myRow` and `myRow~` like the previous line, instead, so that way this will still work, even if we later add additional argument checks to make sure the start sorts before the end.

**Review by ctubbsii — CHANGES_REQUESTED** [review]
This is a good start to check the flush command. However, checking the audit message doesn't guarantee the permission is denied (the auditing could be working, but we still allow the action). So, I think we will need checks to ensure the action is denied. The best place to do that is probably PermissionsIT. Can you make those changes in this PR, or would you prefer to open another PR for PermissionsIT?

File: test/src/main/java/org/apache/accumulo/test/AuditMessageIT.java (hunk 1 above)
Comment (Manno15) [reply test/src/main/java/org/apache/accumulo/test/AuditMessageIT.java:487]: Funny thing is, that is actually what I had originally but I changed it to mimic the parameters names for whatever reason.

**Comment by Manno15** [comment]
I'll take a look at permissionsIT, I can probably just add those to this PR

**Comment by Manno15** [comment]
@ctubbsii As it stands currently, you only need write permissions to flush a table. Do you think I should add separate permissions for Flush specifically (not fully sure how to do this and make it functional. I did add the permission option though), piggyback off the write permission check to also do a flush of the table, or leave as is and assume that permission is checked correctly since they share permissions? 

I also noticed that in `PermissionsIT` that the test case for system permissions isn't actually there and is left as a to-do I assume. Not sure if there was a good reason why that was left undone.

**Comment by ctubbsii** [comment]
> @ctubbsii As it stands currently, you only need write permissions to flush a table. Do you think I should add separate permissions for Flush specifically (not fully sure how to do this and make it functional. I did add the permission option though), piggyback off the write permission check to also do a flush of the table, or leave as is and assume that permission is checked correctly since they share permissions?
> 

There's no need to change any permissions. The test can simply add a user that lacks the appropriate permission, and can verify that the appropriate AccumuloSecurityException denial is thrown if a flush is attempted.

> I also noticed that in `PermissionsIT` that the test case for system permissions isn't actually there and is left as a to-do I assume. Not sure if there was a good reason why that was left undone.

It probably just wasn't in scope of the original issue. Maybe there's another test that checks those? It can be added to this one, if not.

**Comment by Manno15** [comment]
> There's no need to change any permissions. The test can simply add a user that lacks the appropriate permission and can verify that the appropriate AccumuloSecurityException denial is thrown if a flush is attempted.

That's what the test does overall, was more of a question on where to place it. In its own case, or in the Write case. I decided to go with the latter because the former would require adding its own permission.

**Comment by Manno15** [comment]
Whenever I try to grant a system permission on the shell I get: `[shell.Shell] ERROR: org.apache.accumulo.core.util.BadArgumentException: Unrecognized permission near index 6
grant System.SYSTEM -u test_user
`
Am I doing something wrong or is that a bug? This is on 1.10. I am trying to test out the permission for changing system properties.

**Comment by ctubbsii** [comment]
> Whenever I try to grant a system permission on the shell I get: `[shell.Shell] ERROR: org.apache.accumulo.core.util.BadArgumentException: Unrecognized permission near index 6 grant System.SYSTEM -u test_user `
> Am I doing something wrong or is that a bug? This is on 1.10. I am trying to test out the permission for changing system properties.

What command are you running? Compare what you are entering with the help information for the command you are running (`<command> --help`).

**Comment by Manno15** [comment]
Just the Grant command. `grant <perm> -u <user>`. Every other perm works as expected, none of the System ones do for me. It fails on the first S in System.

**Comment by ctubbsii** [comment]
> Just the Grant command. `grant <perm> -u <user>`. Every other perm works as expected, none of the System ones do for me. It fails on the first S in System.

The help documentation for the grant command says to use `-s`. So, `grant -s <systemPerm> -u <user>` should work.

**Comment by Manno15** [comment]
I'll try it tomorrow, but I don't recall seeing that. Maybe I am blind.

**Comment by Manno15** [comment]
@ctubbsii It appears I was blind. The `-s` makes the command work, can't believe is missed it. Any thoughts on my most recent commits and anything else I should add?

**Comment by ctubbsii** [comment]
> @ctubbsii It appears I was blind. The `-s` makes the command work, can't believe is missed it. Any thoughts on my most recent commits and anything else I should add?

The PermissionsIT test is hard to parse. I'm trying to take a look now.

**Comment by ctubbsii** [comment]
The following RPC calls were affected: initiateFlush, waitForFlush, shutdown, shutdownTabletServer, setMasterGoalState, setSystemProperty, and removeSystemProperty. Ideally, we should have test coverage for all of these cases.

**Comment by EdColeman** [comment]
@Manno15 - are you working adding tests for the other calls? Or would it help if someone also took a look?

**Comment by Manno15** [comment]
> @Manno15 - are you working adding tests for the other calls? Or would it help if someone also took a look?

I am splitting time between this and #1824. I am willing to work on adding these other tests but I welcome help with it if you are offering it. 

> The following RPC calls were affected: initiateFlush, waitForFlush, shutdown, shutdownTabletServer, setMasterGoalState, setSystemProperty, and removeSystemProperty. Ideally, we should have test coverage for all of these cases.

Should they be added to this IT? Similar to how I did `flush` or do you think there is a better location for all of these? I am unsure if `permissionsIT` is there to test that the permissions work or there to test the underlying function that uses the permission.

**Comment by dlmarion** [comment]
I'm wondering if a more complete solution here is to create a test for the MasterClientServiceHander (the Master's API) as that is where the SecurityOperation.canFlush method is called.

**Comment by ctubbsii** [comment]
> Should they be added to this IT?

They should be in this PR, but I'm not sure which IT is best suited. I'm sure if you use your best judgment, it will be fine. People can provide help via code reviews if they know of a better place.

**Comment by ctubbsii** [comment]
> I'm wondering if a more complete solution here is to create a test for the MasterClientServiceHander (the Master's API) as that is where the SecurityOperation.canFlush method is called.

Many of those operations already have their own specific end-to-end ITs, as this class represents RPC endpoints. I *thought* the PermissionsIT was intended to specifically check the permissions of each of these endpoints, so that other tests can focus on testing the end-to-end behavior, but I could be wrong about the scope of the PermissionsIT.

**Comment by dlmarion** [comment]
From what I can tell, the driver in PermissionsIT are the different Permissions enum values. I was thinking of a test that hits the API directly, like https://github.com/dlmarion/accumulo/blob/manager-api-it/test/src/main/java/org/apache/accumulo/test/functional/ManagerApiIT.java

**Comment by Manno15** [comment]
I was able to test the flush RPC calls and the set/remove system property call within `PermissionsIT` but I don't see a way to test `shutdown`, `shutdownTabletServer`, and `setMasterGoalState`. I think implementing something akin to what @dlmarion suggested is best suited to handle that so I will implement it into a new IT

**Comment by Manno15** [comment]
I was able to handle each RPC call using @dlmarion's suggestion and using his IT as the basis for the one I pushed. Credit goes to him for that, thanks. The only part I am unsure on is if I should keep it in one test or separate each RPC call to their own test within the IT. 

Overall, I think this is ready for review and we can discuss which approach we want to go with and I can remove what we don't need/want and add anything that is missing.

**Comment by dlmarion** [comment]
LGTM

**Comment by Manno15** [comment]
@dlmarion Do you want me to add you as a co-author if my last commit gets merged? Since it was your idea and you provided the basis for it?

**Comment by dlmarion** [comment]
I'm not really worried about it, up to you.

**Comment by ctubbsii** [comment]
I'm in progress for another review of this. I have some cleanup of the tests to shorten it up a bit.

**Comment by ctubbsii** [comment]
In order to make some improvements to these tests, I backported a JUnit upgrade from #1562 into the 1.10 development branch in d4fd27f32dc2611a23f67b1d3e8dafd8ee05a1cb

**Comment by Manno15** [comment]
I like your changes to the test. I know you already merged but it looks good!
